src/metabase/db/liquibase.clj

```diff
@@ -216,66 +219,92 @@
            " "
            (trs "You can force-release these locks by running `java -jar metabase.jar migrate release-locks`."))))))
     (if (pos? @retry-counter)
-      (log/warnf "Migration lock was cleared after %d retries." @retry-counter)
-      (log/info "No migration lock found."))))
+      (log/warnf "Migration lock was acquired after %d retries." @retry-counter)
+      (do (log/info "No migration lock found.")
+          (log/info "Migration lock acquired.")))))
 
-(defn migrate-up-if-needed!
-  "Run any unrun `liquibase` migrations, if needed."
-  [^Liquibase liquibase ^DataSource data-source]
-  (log/info (trs "Checking if Database has unrun migrations..."))
-  (if (seq (unrun-migrations data-source))
-    (do
-     (log/info (trs "Database has unrun migrations. Checking if migraton lock is taken..."))
-     (wait-for-migration-lock-to-be-cleared liquibase)
-     ;; while we were waiting for the lock, it was possible that another instance finished the migration(s), so make
-     ;; sure something still needs to be done...
-     (let [to-run-migrations      (unrun-migrations data-source)
-           unrun-migrations-count (count to-run-migrations)]
-       (if (pos? unrun-migrations-count)
-         (let [^Contexts contexts nil
-               start-time         (System/currentTimeMillis)]
-           (log/info (trs "Running {0} migrations ..." unrun-migrations-count))
-           (doseq [^ChangeSet change to-run-migrations]
-             (log/tracef "To run migration %s" (.getId change)))
-           (.update liquibase contexts)
-           (log/info (trs "Migration complete in {0}" (u/format-milliseconds (- (System/currentTimeMillis) start-time)))))
-         (log/info
-          (trs "Migration lock cleared, but nothing to do here! Migrations were finished by another instance.")))))
-    (log/info (trs "No unrun migrations found."))))
+(defn- assert-locked [liquibase]
+  (when-not (.hasChangeLogLock (lock-service liquibase))
+    ;; TODO not sure this is the right exception class to use
```

Comment:
Most of the time, we will throw an ExceptionInfo, and I think it's fine here.

Reply:
I wasn't sure what to put in the ex-data in that case, but yeah I think that type is less confusing. I also thought about an IllegalStateException, but ExceptionInfo is pretty ubiquitous for internal application errors.